PROTOCOL.md

```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
+ `groupKeyIds`| `array`   | An array of strings containing the `groupKeyIds` that are being requested.
+ 
+Example of a `GroupKeyRequest` including the rest of the `Stream Layer` fields:
+
+```
+[32, [...msgIdFields], [...msgRefFields], 28, 0, 0, ["requestId", "streamId", "rsaPublicKey", ["keyId"]], 2, "0x29c057786Fa..."]
+```
+
+### GroupKeyResponse
+
+Sent by a publisher as a response to a `GroupKeyRequest` to the requestor's key exchange stream. The requestor should handle this message by adding the contained keys to their key storage, and re-attempting to decrypt any messages that previously could not be decrypted due to missing the keys. 
+
+`GroupKeyResponse`s are always encrypted with (`encryptionType 1 (RSA)`) for the `rsaPublicKey` provided in the `GroupKeyRequest`. The `content` encoded as `contentType 0` (JSON) is as follows: encoded as `contentType 0` (JSON) is as follows:
+
+```
+[requestId, streamId, [groupKeys]]
+```
+
+ Field        | Type      | Description
+ ------------ | --------- | -----------
+ `requestId`  | `string`  | The `requestId` of the `GroupKeyRequest`, identifying which request is being responded to.
+ `streamId`   | `string`  | The stream for which a key is being delivered.
+ `groupKeys`  | `array`   | Array of `[groupKeyId, encryptedGroupKey]` pairs, containing the requested keys encrypted with RSA for the public key provided in the `GroupKeyRequest`. The `encryptedGroupKey` is a hex-encoded binary string.
+ 
+Example of a `GroupKeyResponse` including the rest of the `Stream Layer` fields:
+
+```
+[32, [...msgIdFields], [...msgRefFields], 29, 0, 0, ["requestId", "streamId", ["keyId","123abc"]], 2, "0x29c057786Fa..."]
+```
+
+### GroupKeyReset
+
+Sent by publishers to each subscribers' key exchange streams whenever they want to re-key the stream (in order to revoke access from some subscribers). 
+
+The requestor should handle this message by adding the contained keys to their key storage.
+
+`GroupKeyReset`s are encrypted with (`encryptionType 1 (RSA)`) and are sent to subscribers whose `rsaPublicKey` is known by the publisher (due to receiving a `GroupKeyRequest` earlier, for example. The `content` encoded as `contentType 0` (JSON) is as follows:
+
+```
+[streamId, groupKeyId, groupKey]
+```
+
+ Field        | Type      | Description
+ ------------ | --------- | -----------
+ `streamId`   | `string`  | The stream for which a key is being delivered.
+ `groupKeyId` | `string`  | The id of the `groupKey`.
+ `groupKey`   | `string`  | The group key as a hex-encoded binary string.
+ 
+Example of a `GroupKeyReset` including the rest of the `Stream Layer` fields:
+
+```
+[32, [...msgIdFields], [...msgRefFields], 30, 0, 0, ["keyId", "123abc"], 2, "0x29c057786Fa..."]
+```
+
+### GroupKeyErrorResponse
+
+In case there's an error processing a `GroupKeyRequest`, this message gets sent by a publisher to the requestor's key exchange stream. 
```

Comment:
Is this what would be sent if the `groupKeyId` doesn't exist or the user doesn't have authorisation for that key e.g. user product subscription time expired?